src/include/codegen/interpreter/bytecode_function.h

```diff
@@ -0,0 +1,358 @@
+ * mode). It can be moved and copied.
+ */
+class BytecodeFunction {
+ public:
+  /**
+   * Returns the Opcode enum for a given Opcode Id (to avoid plain casting)
+   * @param id Opcode Id
+   * @return Opcode enum
+   */
+  ALWAYS_INLINE inline static constexpr Opcode GetOpcodeFromId(index_t id) {
+    return static_cast<Opcode>(id);
+  }
+
+  /**
+   * Returns the Opcode Id to a given Opcode enum (to avoid plain casting)
+   * @param opcode Opcode enum
+   * @return Opcode Id
+   */
+  ALWAYS_INLINE inline static constexpr index_t GetOpcodeId(Opcode opcode) {
+    return static_cast<index_t>(opcode);
+  }
+
+  /**
+   * Returns a numan readable string to a given Opcode
+   * @param opcode Opcode enum
+   * @return String representation if the Opcode
+   */
+  static const char *GetOpcodeString(Opcode opcode);
+
+  /**
+   * Returns the overall number of existing Opcodes (not trivial, as the Opcodes
+   * are created with expanding macros)
+   * @return overall number of existing Opcodes
+   */
+  inline static constexpr size_t GetNumberOpcodes() {
+    return static_cast<index_t>(Opcode::NUMBER_OPCODES);
+  }
+
+  /**
+   * Return the instruction pointer to a given instruction index (from this
+   * bytecode function)
+   * @param index instruction index
+   * @return pointer to the instruction at that index inside the bytecode
+   */
+  ALWAYS_INLINE inline const Instruction *GetIPFromIndex(index_t index) const {
+    return reinterpret_cast<const Instruction *>(
+        const_cast<instr_slot_t *>(bytecode_.data()) + index);
+  }
+
+  /**
+   * Returns the instruction index for a given instruction pointer (from this
+   * bytecode function)
+   * @param instruction pointer to a given instruction inside the bytecode
+   * @return index to the instruction the pointer is pointing to
+   */
+  ALWAYS_INLINE inline index_t GetIndexFromIP(
+      const Instruction *instruction) const {
+    index_t index =
+        reinterpret_cast<const instr_slot_t *>(instruction) - bytecode_.data();
+    return index;
+  }
+
+#ifndef NDEBUG
+  const llvm::Instruction *GetIRInstructionFromIP(index_t instr_slot) const;
+#endif
+
+  /**
+   * Returns the number of slots a given instruction occupies in the bytecode
+   * stream.
+   * @param instruction pointer to the instruction inside the bytecode
+   * @return number of slots (each 8 Byte) that are used by this instruction
+   */
+  static size_t GetInstructionSlotSize(const Instruction *instruction);
+
+  /**
+   * Returns the number of slots a given internal call instruction occupies in
+   * the bytecode stream. Internal instructions have a variable length, so the
+   * size has to be calculated.
+   * @param instruction pointer to instruction of type internal call
+   * @return number of slots (each 8 Byte) that are used by this instruction
+   */
+  static ALWAYS_INLINE inline size_t GetInteralCallInstructionSlotSize(
+      const InternalCallInstruction *instruction) {
+    const size_t number_slots =
```

Comment:
Can you briefly explain this calculation and how it ties into the comment for `InternalCallInstruction`?

Reply:
Except for `InternalCallInstruction` all instructions in the bytecode have a fixed size, they always occupy the same number of instruction slots in the bytecode stream. 
But since an internal function call can have an arbitrary number of arguments, the instruction can have different run-time sizes. This number of arguments is encoded in the instruction. 
This function simply reads that number, adds the 4 other arguments and calculates how many instruction slots it occupies. 

![grafik](https://user-images.githubusercontent.com/5031362/41477880-ab7864de-7093-11e8-90dc-2c940fb39dc8.png)


The figure is from my thesis. Do you want me to adapt the comments to make this clearer?

Btw, I changed the calculation to use `MathUtil::DivRoundUp`.

Reply:
I just changed this calculation to a more generic way using `offsetof`:
```
const size_t number_slots = MathUtil::DivRoundUp(offsetof(InternalCallInstruction, args) + sizeof(uint16_t) * instruction->number_args, sizeof(instr_slot_t));
```
This change will be not in this PR, but in the follow-up PR that connects the interpreter to the execution engine.

Reply:
The figure makes sense. Just add a bit of explanation in comments.